src/DotNetty.Buffers/PoolArena.cs

```diff
@@ -0,0 +1,727 @@
+                        return;
+                    }
+                    tableIdx = TinyIdx(normCapacity);
+                    table = this.tinySubpagePools;
+                }
+                else
+                {
+                    if (cache.AllocateSmall(this, buf, reqCapacity, normCapacity))
+                    {
+                        // was able to allocate out of the cache so move on
+                        return;
+                    }
+                    tableIdx = SmallIdx(normCapacity);
+                    table = this.smallSubpagePools;
+                }
+
+                PoolSubpage<T> head = table[tableIdx];
+
+                /**
+                 * Synchronize on the head. This is needed as {@link PoolSubpage#allocate()} and
+                 * {@link PoolSubpage#free(int)} may modify the doubly linked list as well.
+                 */
+                lock (head)
+                {
+                    PoolSubpage<T> s = head.Next;
+                    if (s != head)
+                    {
+                        Contract.Assert(s.DoNotDestroy && s.ElemSize == normCapacity);
+                        long handle = s.Allocate();
+                        Contract.Assert(handle >= 0);
+                        s.Chunk.InitBufWithSubpage(buf, handle, reqCapacity);
+
+                        if (tiny)
+                        {
+                            ++this.allocationsTiny;
+                        }
+                        else
+                        {
+                            ++this.allocationsSmall;
+                        }
+                        return;
+                    }
+                }
+                this.AllocateNormal(buf, reqCapacity, normCapacity);
+                return;
+            }
+            if (normCapacity <= this.ChunkSize)
+            {
+                if (cache.AllocateNormal(this, buf, reqCapacity, normCapacity))
+                {
+                    // was able to allocate out of the cache so move on
+                    return;
+                }
+                this.AllocateNormal(buf, reqCapacity, normCapacity);
+            }
+            else
+            {
+                // Huge allocations are never served via the cache so just call allocateHuge
+                this.AllocateHuge(buf, reqCapacity);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.Synchronized)]
+        void AllocateNormal(PooledByteBuffer<T> buf, int reqCapacity, int normCapacity)
+        {
+            ++this.allocationsNormal;
+            if (this.q050.Allocate(buf, reqCapacity, normCapacity) || this.q025.Allocate(buf, reqCapacity, normCapacity)
+                || this.q000.Allocate(buf, reqCapacity, normCapacity) || this.qInit.Allocate(buf, reqCapacity, normCapacity)
+                || this.q075.Allocate(buf, reqCapacity, normCapacity) || this.q100.Allocate(buf, reqCapacity, normCapacity))
```

Comment:
The netty version of this function doesn't appear to allocate using the `q100` chunk list. I'm not sure if that is a bug in netty or an oversight here.